DevGuide lacks section on interoperability (#4984)
OpenDDS/OpenDDS: docs/devguide/introduction.rst

  * The rest of the building block is not supported.

*********************************************
Building and Configuring for Interoperability
*********************************************

> The two components needed for interoperability are the RTPS Discovery library (e.g., ``libOpenDDS_Rtps``) and the RTPS UDP transport library (e.g., ``libOpenDDS_Rtps_Udp``).

Reviewer: ```suggestion
The two components needed for interoperability are the :ref:`RTPS Discovery library <rtps-disc>` and the :ref:`RTPS UDP transport library <rtps-udp-transport>`.
```

These sections have the filenames as well as other info.

Both of these libraries are built and installed by default.
Those using a custom build system need to make sure that these libraries are available to the application.

As of OpenDDS 3, these components will not be used by default and must be configured.
This can be done via :ref:`environment variables, command-line options, configuration files, and/or code <config-store-keys>`.

The following configuration file snippet changes the default discovery to RTPS and the default transport to RTPS UDP.

.. code-block:: ini

    [common]
    DCPSDefaultDiscovery=DEFAULT_RTPS
    DCPSGlobalTransportConfig=$file

    [transport/the_rtps_transport]
    transport_type=rtps_udp

The following code accomplishes the same:

.. code-block:: cpp

    using namespace OpenDDS::DCPS;
    using namespace OpenDDS::RTPS;
    TransportConfig_rch config =

>       TransportRegistry::instance()->create_config("rtps_interop_demo");

Reviewer: Maybe some other name?

    TransportInst_rch inst =
      TransportRegistry::instance()->create_inst("the_rtps_transport", "rtps_udp");
    config->instances_.push_back(inst);
    TransportRegistry::instance()->global_config(config);

    RtpsDiscovery_rch disc = make_rch<RtpsDiscovery>("RtpsDiscovery");

>     disc->use_xtypes(RtpsDiscoveryConfig::XTYPES_NONE);

Reviewer 2: The default value of `use_xtypes` is minimal (see `bool RtpsDiscoveryConfig::use_xtypes() const`), so this should change to match what the ini does.  Also the devguide page that lists these parameters may need to be updated.

Author: Should we 1) remove `use_xtypes` from the example since minimal is the default or 2) expand the example with `UseXTypes` to suggest how changing to complete might be done?

Reviewer 2: I think we should go with `2)`, it's a good starting point for the audience of this section of the docs.

Reviewer 2: @jrw972 looks like this one is not resolved?

    TheServiceParticipant->add_discovery(static_rchandle_cast<Discovery>(disc));
    TheServiceParticipant->set_default_discovery(disc->key());

The remaining concern for interoperability is the use of :ref:`xtypes` features.
See :ref:`xtypes--unimplemented-features` and :ref:`xtypes--differences-from-the-specification`.
If XTypes is not being used, then see :ref:`xtypes--interoperability-with-non-xtypes-implementations`.

For maximum interoperability, complete type objects should be used.
See :ref:`xtypes--enabling-use-of-completetypeobjects`.

:ref:`opendds_idl` has options for changing the defaults when processing types.
Using these options may reduce the number of annotations needed in IDL.
See :option:`opendds_idl --default-nested`, :option:`opendds_idl --default-extensibility`, :option:`opendds_idl --default-autoid`, and :option:`opendds_idl --default-try-construct`.

> For maximum interoperability, the :ref:`xtypes--xcdr2` data representation should be used.

Reviewer 2: Suggest reworking this:

XTypes Data Representation is ... (give a little background, link to existing docs especially about limitations with XCDR1)

* OpenDDS Data Readers default to supporting both XCDR1 and XCDR2

* OpenDDS Data Writers default to producing XCDR2.  Based on other DDS implementations which will participate in interoperability, determine if XCDR2 or XCDR1 is appropriate for your application.

OpenDDS has limited support for XCDR1 (see :ref:`xtypes--xcdr1-support`).
XCDR2 is the default.

> It can be set explicitly via QoS or by using the non-standard :ref:`xtypes--anno-opendds-data-representation-xcdr2` annotation.

Reviewer: Annotations don't set the encodings, they only restrict what encodings a reader or writer is allowed to successfully start with:

https://github.com/OpenDDS/OpenDDS/blob/9bcc5b2b6e2fbdfb5d445b2e1be168d948d53477/docs/devguide/xtypes.rst?plain=1#L567-L568

[A combination of QoS and the transport](https://opendds.readthedocs.io/en/master/devguide/quality_of_service.html#data-representation-qos) is the only thing that defines encoding.

The following snippet shows how to choose XCDR2 for a DataWriter:

.. code-block:: cpp

    DDS::DataWriterQos qos;
    pub->get_default_datawriter_qos(qos);
    qos.representation.value.length(1);

>     qos.representation.value[0] = DDS::XCDR2_DATA_REPRESENTATION;

Reviewer: This seems kinda redundant with this in the xtypes chapter:

https://github.com/OpenDDS/OpenDDS/blob/9bcc5b2b6e2fbdfb5d445b2e1be168d948d53477/docs/devguide/xtypes.rst?plain=1#L451-L458

> This setting must be allowed by the transport (which must be RTPS/UDP for interoperability).

Reviewer 2: ```suggestion
```
We could remove this line since the transport is already discussed above.

The following code accomplishes the same:

.. code-block:: cpp

    using namespace OpenDDS::DCPS;
    using namespace OpenDDS::RTPS;
    TransportConfig_rch config =
      TransportRegistry::instance()->create_config("rtps_config");
    TransportInst_rch inst =
      TransportRegistry::instance()->create_inst("the_rtps_transport", "rtps_udp");
    config->instances_.push_back(inst);
    TransportRegistry::instance()->global_config(config);

    RtpsDiscovery_rch disc = make_rch<RtpsDiscovery>("RtpsDiscovery");
    disc->use_xtypes(RtpsDiscoveryConfig::XTYPES_COMPLETE);
    TheServiceParticipant->add_discovery(static_rchandle_cast<Discovery>(disc));
    TheServiceParticipant->set_default_discovery(disc->key());

Basic interoperability in DDS requires 1) all participants are in the same domain, 2) DataWriters and DataReaders must agree on the name of Topics, and 3) the QoS for DataWriters and DataReaders must be compatible.

Security is another dimension of interoperability.
Participants must agree on the identity and permissions certificate authorities, governance, and have sufficient permissions.
See :ref:`dds_security` for the details of using DDS Security with OpenDDS.

> The remaining concern for interoperability are the use of :ref:`xtypes` features.

Reviewer 2: ```suggestion
The remaining concern for interoperability is the use of :ref:`xtypes` features.
```

XTypes :ref:`xtypes--data-representation` determines the format used by DataWriters to encode samples and the set of formats that can be used by a DataReader to decode samples.
The set of formats for a DataReader must include the formats used by the DataWriters for interoperability.
The two interoperable formats supported by OpenDDS are :ref:`xtypes--xcdr1` and :ref:`xtypes--xcdr2`.
OpenDDS has limited support for XCDR1 (see :ref:`xtypes--xcdr1-support`).
OpenDDS DataReaders default to XCDR1 and XCDR2 and OpenDDS DataWriters default to XCDR2.
For interoperability, the data representation should be considered and adjusted appropriately.

The data representation can be changed using the :ref:`qos-data-representation`.
The non-standard :ref:`xtypes--anno-opendds-data-representation-xcdr1` and :ref:`xtypes--anno-opendds-data-representation-xcdr2` annotations can be used to limit the values that can be used in the :ref:`qos-data-representation`.
See :ref:`xtypes--data-representation` for an example of setting the :ref:`qos-data-representation`.

Failures in interoperability are often silent.

> A good first step is to enable :ref:`logging` and check for errors and warnings.

Reviewer: ```suggestion
A good first step is to enable :ref:`run_time_configuration--logging` and check for errors and warnings.
```

The DDS API can assist in debugging but often lacks the detail to correct issues.
See :ref:`built_in_topics` and :ref:`conditions_and_listeners` for details.
If logging and introspection is not enough, then the next step is usually a packet capture and analysis, e.g., Wireshark.
The capture should show the following sequence:

> 1. Participants discoverying one another.

Reviewer 2: ```suggestion
1. Participants discovering one another.
```
